File: src/test/java/kitchenpos/domain/FakeMenuRepository.java
package kitchenpos.domain;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

public class FakeMenuRepository implements MenuRepository {

    private final HashMap<UUID, Menu> inMemory = new HashMap<>();

    @Override
    public List<Menu> findAllByIdIn(List<UUID> ids) {
        return ids.stream()
                .filter(inMemory::containsKey)
                .map(inMemory::get)
                .toList();
    }

    @Override
    public List<Menu> findAllByProductId(UUID productId) {
        return inMemory.values()
                .stream()
                .filter(menu -> menu
                        .getMenuProducts()
                        .stream()
                        .anyMatch(menuProduct
                                -> menuProduct
                                .getProduct()
                                .getId().equals(productId)))
                .toList();
    }

    @Override
    public Menu save(Menu menu) {
        inMemory.put(menu.getId(), menu);
        return menu;
    }

    @Override
    public Optional<Menu> findById(UUID id) {
        return Optional.ofNullable(inMemory.get(id));
    }

    @Override
    public List<Menu> findAll() {
        return new ArrayList<>(inMemory.values());
    }
}

File: src/test/java/kitchenpos/domain/FakeMenuRepository.kt
package kitchenpos.domain

import java.util.Optional
import java.util.UUID

class FakeMenuRepository : InMemoryRepository<Menu>(), MenuRepository {
    override fun findAllByIdIn(ids: MutableList<UUID>?): MutableList<Menu> {
        return items.filter { it.id in (ids ?: emptyList()) }.toMutableList()
    }

    override fun findAllByProductId(productId: UUID?): MutableList<Menu> {
        return items.filter { menu -> menu.menuProducts.any { it.productId == productId } }.toMutableList()
    }

    override fun findById(id: UUID): Optional<Menu>? {
        return items.find { it.id == id }?.let { Optional.of(it) } ?: Optional.empty()
    }
}
